### PR TITLE
[SPARK-30036][SQL] Fix: REPARTITION hint does not work with order by

### PR DESCRIPTION
**Pull request by jackylee-ch**
### Why are the changes needed?
`EnsureRequirements` adds `ShuffleExchangeExec` (RangePartitioning) after Sort if `RoundRobinPartitioning` behinds it. This will cause 2 shuffles, and the number of partitions in the final stage is not the number specified by `RoundRobinPartitioning.

**Example SQL**
```
SELECT /*+ REPARTITION(5) */ * FROM test ORDER BY a
```

**BEFORE**
```
== Physical Plan ==
*(1) Sort [a#0 ASC NULLS FIRST], true, 0
+- Exchange rangepartitioning(a#0 ASC NULLS FIRST, 200), true, [id=#11]
   +- Exchange RoundRobinPartitioning(5), false, [id=#9]
      +- Scan hive default.test [a#0, b#1], HiveTableRelation `default`.`test`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [a#0, b#1]
```

**AFTER**
```
== Physical Plan ==
*(1) Sort [a#0 ASC NULLS FIRST], true, 0
+- Exchange rangepartitioning(a#0 ASC NULLS FIRST, 5), true, [id=#11]
   +- Scan hive default.test [a#0, b#1], HiveTableRelation `default`.`test`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [a#0, b#1]
```

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Run suite Tests and add new test for this.

### CODE REVIEW

**Comment by stczwd** [comment]
cc @HyukjinKwon @cloud-fan @xuanyuanking

**Comment by ulysses-you** [comment]
I think we should add check at `CollapseRepartition` instead of `EnsureRequirements` if we want to prune repartition with sort.

**Comment by cloud-fan** [comment]
ok to test

**Comment by stczwd** [comment]
> I think we should add check at `CollapseRepartition` instead of `EnsureRequirements` if we want to prune repartition with sort.

Adding  RangePartitioning doesn't happen in optimizer, thus we can't check this in `CollapseRepartition`. Besides, It is not easy to keep partition number after pruning repartition with sort.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 1]
```diff
@@ -55,6 +55,10 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
         child
       case (child, BroadcastDistribution(mode)) =>
         BroadcastExchangeExec(mode, child)
+      case (ShuffleExchangeExec(partitioning: RoundRobinPartitioning, child, _),
+          distribution: OrderedDistribution) =>
+        ShuffleExchangeExec(
+          distribution.createPartitioning(partitioning.numPartitions), child)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: Could you update like the following, @stczwd ?
```scala
-      case (ShuffleExchangeExec(partitioning: RoundRobinPartitioning, child, _),
-          distribution: OrderedDistribution) =>
-        ShuffleExchangeExec(
-          distribution.createPartitioning(partitioning.numPartitions), child)
+      case (ShuffleExchangeExec(partitioning, child, _), distribution) =>
+        ShuffleExchangeExec(distribution.createPartitioning(partitioning.numPartitions), child)
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 2]
```diff
@@ -421,6 +421,24 @@ class PlannerSuite extends SharedSparkSession {
...
+    assert(!partitioning.satisfies(distribution))
+
+    val inputPlan = SortExec(SortOrder(Literal(1), Ascending) :: Nil,
+      global = true,
+      child = ShuffleExchangeExec(
+        partitioning,
+        DummySparkPlan(outputPartitioning = partitioning)))
+    val outputPlan = EnsureRequirements(spark.sessionState.conf).apply(inputPlan)
+    assert(outputPlan.find{
+      case e: ShuffleExchangeExec => e.outputPartitioning.isInstanceOf[RoundRobinPartitioning]
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]:
```scala
- case e: ShuffleExchangeExec => e.outputPartitioning.isInstanceOf[RoundRobinPartitioning]
+ case ShuffleExchangeExec(_: RoundRobinPartitioning, _, _) => true
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 3]
```diff
@@ -421,6 +421,24 @@ class PlannerSuite extends SharedSparkSession {
...
+    val partitioning = RoundRobinPartitioning(5)
+    assert(!partitioning.satisfies(distribution))
+
+    val inputPlan = SortExec(SortOrder(Literal(1), Ascending) :: Nil,
+      global = true,
+      child = ShuffleExchangeExec(
+        partitioning,
+        DummySparkPlan(outputPartitioning = partitioning)))
+    val outputPlan = EnsureRequirements(spark.sessionState.conf).apply(inputPlan)
+    assert(outputPlan.find{
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: `.find{` -> `.find {`.

**Comment by dongjoon-hyun** [comment]
I updated the PR description a little, @stczwd .

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 4]
```diff
@@ -55,6 +55,10 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
         child
       case (child, BroadcastDistribution(mode)) =>
         BroadcastExchangeExec(mode, child)
+      case (ShuffleExchangeExec(partitioning: RoundRobinPartitioning, child, _),
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: How about use `Partitioning` instead of `RoundRobinPartitioning `. Since we already support this `SELECT /*+ REPARTITION(5, a) */ * FROM test ORDER BY a`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 5]
```diff
@@ -421,6 +421,24 @@ class PlannerSuite extends SharedSparkSession {
     }
   }
 
+  test("SPARK-30036: EnsureRequirements replace Exchange " +
+      "if child has SortExec and RoundRobinPartitioning") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: How about just saying `Remove unnecessary RoundRobinPartitioning` in the test title? Also, can you make the PR title obivious, too?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala (hunk 1 above)
Comment (stczwd) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: You means this should work in other Partitioning? Let me run some test for it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 2 above)
Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: good

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala (hunk 4 above)
Comment (stczwd) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: thanks, I will change it

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 3 above)
Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: done

**Comment by stczwd** [comment]
> I updated the PR description a little, @stczwd .

@dongjoon-hyun  thanks

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 5 above)
Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: Good for test title, thanks.
But it is not suitable for PR title, there are other situations in this titile.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala (hunk 1 above)
Comment (stczwd) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: thanks, i have change my code

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 5 above)
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: How about `Avoid RoundRobinPartitioning that EnsureRequirements Redundantly adds`?

File: sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala [hunk 6]
```diff
@@ -39,9 +39,8 @@ class ConfigBehaviorSuite extends QueryTest with SharedSparkSession {
     def computeChiSquareTest(): Double = {
       val n = 10000
       // Trigger a sort
-      // Range has range partitioning in its output now. To have a range shuffle, we
-      // need to run a repartition first.
-      val data = spark.range(0, n, 1, 1).repartition(10).sort($"id".desc)
+      // Range has range partitioning in its output now.
+      val data = spark.range(0, n, 1, 10).sort($"id".desc)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: why this change?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 7]
```diff
@@ -421,6 +421,42 @@ class PlannerSuite extends SharedSparkSession {
...
+
+    val inputPlan = SortExec(SortOrder(Literal(1), Ascending) :: Nil,
+      global = true,
+      child = ShuffleExchangeExec(
+        partitioning,
+        DummySparkPlan(outputPartitioning = partitioning)))
+    val outputPlan = EnsureRequirements(spark.sessionState.conf).apply(inputPlan)
+    assert(outputPlan.find {
+      case ShuffleExchangeExec(_: RoundRobinPartitioning, _, _) => true
+      case _ => false}.isEmpty,
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: nit:
```
...find {
  case ...
  case ...
}.isEmpty
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 8]
```diff
@@ -421,6 +421,42 @@ class PlannerSuite extends SharedSparkSession {
...
+
+    val inputPlan = SortExec(SortOrder(Literal(1), Ascending) :: Nil,
+      global = true,
+      child = ShuffleExchangeExec(
+        partitioning,
+        DummySparkPlan(outputPartitioning = partitioning)))
+    val outputPlan = EnsureRequirements(spark.sessionState.conf).apply(inputPlan)
+    assert(outputPlan.find {
+      case ShuffleExchangeExec(_: HashPartitioning, _, _) => true
+      case _ => false}.isEmpty,
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: ditto

**Comment by cloud-fan** [comment]
shall we add an end-to-end test for `SELECT /*+ REPARTITION(5) */ * FROM test ORDER BY a`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 5 above)
Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: Because HashPartitioning should also be concerned.

File: sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala (hunk 6 above)
Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: This test is failed because of pruning repartition, the partition number is not right in this case.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: The test wants to test a sort with shuffle. Do we still have shuffle in this query?

**Comment by stczwd** [comment]
> shall we add an end-to-end test for `SELECT /*+ REPARTITION(5) */ * FROM test ORDER BY a`?
@cloud-fan The modification of this patch only affects the number of final partitions and does not affect the overall result. Is it enough to check the rule? Or, maybe check the number of partitions?

File: sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala (hunk 6 above)
Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: Yes

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 9]
```diff
@@ -55,6 +55,8 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
         child
       case (child, BroadcastDistribution(mode)) =>
         BroadcastExchangeExec(mode, child)
+      case (ShuffleExchangeExec(partitioning, child, _), distribution: OrderedDistribution) =>
+        ShuffleExchangeExec(distribution.createPartitioning(partitioning.numPartitions), child)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: This considers a special case for OrderedDistribution. Generally, if ShuffleExchangeExec is followed by any unsatisfying distribution , we should always trim the ShuffleExchangeExec and apply the partitioning of distribution. Don't we?

Comment (stczwd) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: Sound reasonable. Any suitable cases?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: I just tried few possible cases, but can not have a concrete case like this. Maybe this is the only case possibly.

**Comment by cloud-fan** [comment]
We can add an end-to-end test, check the physical plan of a query, and count shuffles.

File: sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: I'm a bit confused. `spark.range` reports `RangePartitioning`, so there shouldn't be any shuffles. What gets changed?

Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]:
```
== Optimized Logical Plan ==
Sort [id#8L DESC NULLS LAST], true
+- Range (0, 100, step=1, splits=Some(10))

== Physical Plan ==
*(2) Sort [id#8L DESC NULLS LAST], true, 0
+- Exchange rangepartitioning(id#8L DESC NULLS LAST, 4), true, [id=#33]
   +- *(1) Range (0, 100, step=1, splits=10)
```
The sort api will add global=true in SortExec, similar with orderby, which will cause SortExec's requiredChildDistribution to be parsed into OrderedDistribution.
EnsureRequirements will add a ShuffleExchangeExec with rangePartitioning after OrderedDistribution.

**Comment by stczwd** [comment]
> We can add an end-to-end test, check the physical plan of a query, and count shuffles.
Sure，I will add some tests for these cases.

File: sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: In the current master
```
scala> spark.range(0, 10000, 1, 10).sort("id").explain(true)
== Parsed Logical Plan ==
'Sort ['id ASC NULLS FIRST], true
+- Range (0, 10000, step=1, splits=Some(10))

== Analyzed Logical Plan ==
id: bigint
Sort [id#8L ASC NULLS FIRST], true
+- Range (0, 10000, step=1, splits=Some(10))

== Optimized Logical Plan ==
Range (0, 10000, step=1, splits=Some(10))

== Physical Plan ==
*(1) Range (0, 10000, step=1, splits=10)
```

If we add a shuffle now, then it's a regression and we should fix.

File: sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala [hunk 10]
```diff
@@ -39,9 +39,8 @@ class ConfigBehaviorSuite extends QueryTest with SharedSparkSession {
     def computeChiSquareTest(): Double = {
       val n = 10000
       // Trigger a sort
-      // Range has range partitioning in its output now. To have a range shuffle, we
-      // need to run a repartition first.
-      val data = spark.range(0, n, 1, 1).repartition(10).sort($"id".desc)
+      // Range has range partitioning in its output now.
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: shall we remove this comment now? it's not useful anymore as we do add shuffle, the range output partitioning doesn't matter.

File: sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala [hunk 11]
```diff
@@ -55,12 +54,12 @@ class ConfigBehaviorSuite extends QueryTest with SharedSparkSession {
 
     withSQLConf(SQLConf.SHUFFLE_PARTITIONS.key -> numPartitions.toString) {
       // The default chi-sq value should be low
-      assert(computeChiSquareTest() < 100)
+      assert(computeChiSquareTest() < 10)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: the physical plan is same as before, what caused this change?

File: sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala (hunk 10 above)
Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: okey

File: sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala (hunk 11 above)
Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: They are not same, we had two shuffles before, one was RoundRobinPartitioning, the other was RangePartitioning.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala]: ah i see

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 12]
```diff
@@ -421,6 +421,52 @@ class PlannerSuite extends SharedSparkSession {
...
+        partitioning,
+        DummySparkPlan(outputPartitioning = partitioning)))
+    val outputPlan = EnsureRequirements(spark.sessionState.conf).apply(inputPlan)
+    assert(outputPlan.find {
+      case ShuffleExchangeExec(_: HashPartitioning, _, _) => true
+      case _ => false
+    }.isEmpty,
+      "HashPartitioning should be changed to RangePartitioning")
+
+    val query = testData.select('key, 'value).repartition(5, 'key).sort('key.asc)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: I'm not very sure about this. `df.repartition` is a low-level API that allows users to hash-partition the data. There is also a `df.repartitionByRange` to do range partitioning. We shouldn't break users' expectations.

Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: HashPartitioning or RangePartitioning won't take effect if it is followed by OrderedDistribution. If user really want hashPartitioning, they should use sortWithinPartitions.

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 12 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: makes sense

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 13]
```diff
@@ -421,6 +421,52 @@ class PlannerSuite extends SharedSparkSession {
     }
   }
 
+  test("SPARK-30036: Romove unnecessary RoundRobinPartitioning " +
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: nit `Romove` -> `Remove`

Comment (stczwd) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: done

**Comment by HyukjinKwon** [comment]
Looks fine to me

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master, I guess :-).

**Comment by cloud-fan** [comment]
yea merged to master!

**Comment by ulysses-you** [comment]
Wait, another point. First this scence also exists  in `join` or `window` operator, as @viirya say, exists other Distribution. 
For e.g. join
```
val df = spark.range(1, 10, 2)
df.join(df.repartition(10), Seq("id"), "left").explain(true)

// physical plan like this
= Physical Plan ==
*(5) Project [id#0L]
+- SortMergeJoin [id#0L], [id#83L], LeftOuter
   :- *(2) Sort [id#0L ASC NULLS FIRST], false, 0
   :  +- Exchange hashpartitioning(id#0L, 200), true, [id=#378]
   :     +- *(1) Range (1, 10, step=1, splits=40)
   +- *(4) Sort [id#83L ASC NULLS FIRST], false, 0
      +- Exchange hashpartitioning(id#83L, 200), true, [id=#384]
         +- Exchange RoundRobinPartitioning(10), false, [id=#383]
            +- *(3) Range (1, 10, step=1, splits=40)
```

And then there is a little difference between `2 -> 10 -> 200` and `2 -> 10` because of different operator complexity. Repartition may be is a light operator compare with sort or join or else algorithm. So it's not sure `2 -> 10` is always run faster than `2 -> 10 -> 200`.

The last, if end user really want result partition is 10, should use `df.sort("id").repartition(10)` instead, not the `df.repartition(10).sort("id")`. Pruning shuffle may mislead user.

cc @HyukjinKwon @cloud-fan @maropu

**Comment by cloud-fan** [comment]
for join, it doesn't require `OrderedDistribution`, but `HashClusteredDistribution`.

This PR only affects sort.

**Comment by stczwd** [comment]
>  if end user really want result partition is 10, should use df.sort("id").repartition(10) instead, not the df.repartition(10).sort("id"). Pruning shuffle may mislead user.

df.sort("id").repartition(10) returns wrong result. Global sort result would be repartitioned with disordered.

**Comment by ulysses-you** [comment]
> This PR only affects sort.

Yes it is. But it is similar with outer join.  `df.join(df.repartition(10), Seq("id"), "left")` result the 200 partitions, and now `df.repartition(10).sort("id")` result the 10 partitions. Should they be same ?

**Comment by ulysses-you** [comment]
> df.sort("id").repartition(10) returns wrong result. Global sort result would be repartitioned with disordered.

Sorry for the wrong example. I mean user should use the right way to change partition. Obviously `df.repartition(10).sort("id")` should return the spark sql shuffle partitions.

**Comment by stczwd** [comment]
> Yes it is. But it is similar with outer join. `df.join(df.repartition(10), Seq("id"), "left")` result the 200 partitions, and now `df.repartition(10).sort("id")` result the 10 partitions. Should they be same ?
> Sorry for the wrong example. I mean user should use the right way to change partition. Obviously df.repartition(10).sort("id") should return the spark sql shuffle partitions.

Thanks for pay attention on this. The main problem you described is whether we should change partition num for OrderedDistribution. 
Hm, it's you add `REPARTITION` hint in [SPARK-28746](https://github.com/apache/spark/pull/25464), you may know what it means to users. In other cases, `REPARTITION` hint will change result partition number with shuffles, but it didn't work with order by, which confused users. `REPARTITION` is a great way for users to control final result num, we should keep it works on every queries.
Besides, `sort("id")` is a global OrderDistribution, which usually generate the final result. It is not easy to set partition number with defaultShufflePartitions, especially on large queries with multiple shuffles. 
Finally, changing partition num is good way for use to control shuffle and final results with `df.repartition(10).sort("id")`. Users may won't write `df.repartition(10).sort("id")` unless they want change the final partition num. It is not a normal case in other scenes.

Correct me if I'm wrong. Thanks

**Comment by ulysses-you** [comment]
I see you want to make a way that change partition easily after sort.

Only one thing I not sure. If `df.repartition(10).sort("id")` result the 10 partitions, user will think `df.repartition(10).xxx` result 10 partitions too but actually not. It's a special handle for sort.

I don not know how committer think about it, or it's just fine.

**Comment by cloud-fan** [comment]
what `.sort` should guarantee is that the output is ordered, and users shouldn't care about the number of partitions.

It's more efficient to shuffle only once for query `df.repartition(10).sort("id")`. This is just an optimization and nothing about semantic.

For `df.join(df.repartition(10), Seq("id"), "left")`, again we don't care about the number of result partitions. If there is a way to save shuffles, please propose.

**Comment by maryannxue** [comment]
Think we should revert this PR. The change in test `ConfigBehaviorSuite` cannot be fully justified. Plus, this is not the right approach to fix this kind of issue.

**Comment by cloud-fan** [comment]
After more thoughts, I think it's wrong to use optimization to fix a bug.

Looking into the bug, the issue is: the `Repartition` operator added by the hint is under the `Sort` operator, not above it. This is because our parser treats ORDER BY as the last clause, while the hint is associated with the SELECT clause. The parser rule is like `SELECT ... UNION/INTERSECT SELECT ... ORDER BY`. That's why we add the `Sort` operator at the end.

I think #27096 is in the right way to optimize redundant shuffles, but we still need to fix the bug about how to handle hints in the parser.

I'm reverting this. Let's fix the bug in the parser.
